Add Create, Control, Observe introduction guides (#370)
onkernel/docs: introduction/observe.mdx

---
title: "Observe"
description: "Watch your agent work, debug what went wrong"
---

Browser agents fail in ways that don't show up in logs. Kernel gives you four ways to see what's actually happening — live, after the fact, frame by frame, and line by line.

## Live view

Every browser exposes a `browser_live_view_url` you can open in a browser tab or embed in an iframe. Use it to watch an agent run in real time, hand a session off to a human-in-the-loop, or surface the browser as part of your own UI.

<CodeGroup>
```typescript Typescript/Javascript
import Kernel from '@onkernel/sdk';

const kernel = new Kernel();

const kernelBrowser = await kernel.browsers.create();
console.log(kernelBrowser.browser_live_view_url);
```

```python Python
from kernel import Kernel

kernel = Kernel()

kernel_browser = kernel.browsers.create()
print(kernel_browser.browser_live_view_url)
```

```bash CLI
kernel browsers view <session-id>
```
</CodeGroup>

> Add `?readOnly=true` for a non-interactive view, or enable [kiosk mode](/browsers/live-view#kiosk-mode) at creation for a fullscreen, chromeless experience. Full reference: [Live View](/browsers/live-view).

Reviewer: ```suggestion
Add `?readOnly=true` for a non-interactive view, or enable [kiosk mode](/browsers/live-view#kiosk-mode) at creation for a fullscreen, cinematic experience. Full reference: [Live View](/browsers/live-view).
```

## Replays

> Replays are MP4 recordings you start and stop on demand. They're the right tool for post-hoc debugging: a failed run produces a video you can scrub through, share, or attach to a bug report.

Reviewer: ```suggestion
Replays are MP4 recordings you start and stop on demand — capture as many clips per session as you need. They're the right tool for post-hoc debugging: a failed run gives you one or more videos to scrub through, share, or attach to a bug report.
```

> 

Reviewer: ```suggestion

Replays can also be enabled on managed auth sessions, so you can [debug failed logins](https://www.kernel.sh/docs/auth/configuration#record-sessions-for-debugging) the same way.

```
